File: frontend/src/api/knowledge.ts
import axios from 'axios'

const api = axios.create({ baseURL: '/api', headers: { 'Content-Type': 'application/json' } })

export interface KnowledgeItem {
  id: string
  category: string
  subType: string
  title: string
  content: string
  tags: string[]
  isBuiltin: boolean
  createdAt: string
  updatedAt: string
}

export const CATEGORIES = [
  { value: 'plot', label: '情节题材', desc: '宗门大比、秘境探险、复仇等' },
  { value: 'character', label: '人物原型', desc: '废柴逆袭、天才型、反派模板' },
  { value: 'world', label: '世界设定', desc: '修炼体系、宗门、天材地宝' },
  { value: 'satisfying', label: '爽点模板', desc: '打脸套路、突破场景、系统奖励' },
  { value: 'foreshadow', label: '伏笔套路', desc: '伏笔类型、回收方式' },
  { value: 'chapter_template', label: '章节模板', desc: '战斗章、修炼章、剧情章' },
  { value: 'writing_technique', label: '写作技巧', desc: '对话、描写、视角、节奏' },
]

export async function listKnowledge(category?: string, search?: string): Promise<KnowledgeItem[]> {
  const params: Record<string, string> = {}
  if (category) params.category = category
  if (search) params.search = search
  const { data } = await api.get('/knowledge', { params })
  return data
}

export async function getKnowledge(id: string): Promise<KnowledgeItem> {
  const { data } = await api.get(`/knowledge/${id}`)
  return data
}

export async function createKnowledge(item: {
  category: string; subType?: string; title: string; content: string; tags?: string[]
}): Promise<KnowledgeItem> {
  const { data } = await api.post('/knowledge', item)
  return data
}

export async function updateKnowledge(id: string, item: Partial<KnowledgeItem>): Promise<KnowledgeItem> {
  const { data } = await api.put(`/knowledge/${id}`, item)
  return data
}

export async function deleteKnowledge(id: string): Promise<void> {
  await api.delete(`/knowledge/${id}`)
}

// ===== 角色预设选项 =====

export interface PresetGroup {
  title: string
  options: string[]
}

export async function getCharacterPresets(subType: string): Promise<PresetGroup[]> {
  const { data } = await api.get(`/knowledge/presets/${subType}`)
  return data
}

File: backend/app/api/knowledge.py
"""知识库 API 路由"""

import json
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel
from typing import Optional

from app.services import knowledge_service

router = APIRouter(prefix="/api/knowledge", tags=["knowledge"])


class KnowledgeCreateRequest(BaseModel):
    category: str
    subType: Optional[str] = ""
    title: str
    content: str
    tags: Optional[list[str]] = []


class KnowledgeUpdateRequest(BaseModel):
    category: Optional[str] = None
    subType: Optional[str] = None
    title: Optional[str] = None
    content: Optional[str] = None
    tags: Optional[list[str]] = None


@router.get("")
async def list_knowledge(
    category: Optional[str] = Query(None),
    search: Optional[str] = Query(None),
    sub_type: Optional[str] = Query(None),
):
    return await knowledge_service.list_knowledge(category=category, sub_type=sub_type, search=search)


@router.get("/presets/{sub_type}")
async def get_character_presets(sub_type: str):
    """返回角色预设选项，按标题分组，content 为 JSON 数组"""
    items = await knowledge_service.list_knowledge(
        category="character_preset", sub_type=sub_type
    )
    result = []
    for item in items:
        try:
            options = json.loads(item["content"])
        except (json.JSONDecodeError, TypeError):
            options = []
        result.append({"title": item["title"], "options": options})
    return result


@router.get("/{item_id}")
async def get_knowledge(item_id: str):
    item = await knowledge_service.get_knowledge(item_id)
    if not item:
        raise HTTPException(404, "Knowledge item not found")
    return item


@router.post("", status_code=201)
async def create_knowledge(req: KnowledgeCreateRequest):
    return await knowledge_service.create_knowledge(req.model_dump())


@router.put("/{item_id}")
async def update_knowledge(item_id: str, req: KnowledgeUpdateRequest):
    data = {k: v for k, v in req.model_dump().items() if v is not None}
    item = await knowledge_service.update_knowledge(item_id, data)
    if not item:
        raise HTTPException(404, "Knowledge item not found")
    return item


@router.delete("/{item_id}", status_code=204)
async def delete_knowledge(item_id: str):
    deleted = await knowledge_service.delete_knowledge(item_id)
    if not deleted:
        raise HTTPException(404, "Item not found or is built-in")
